core_types/src/dna/fn_declarations.rs

```diff
@@ -0,0 +1,115 @@
+//! File holding all the structs for handling function declarations defined in DNA.
+
+use crate::dna::capabilities::CapabilityType;
+
+/// Represents the type declaration for zome function parameter
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Hash)]
+pub struct FnParameter {
+    #[serde(rename = "type")]
+    pub parameter_type: String,
+    pub name: String,
+}
+
+impl FnParameter {
+    #[allow(dead_code)]
+    pub fn new<S: Into<String>>(n: S, t: S) -> FnParameter {
+        FnParameter {
+            name: n.into(),
+            parameter_type: t.into(),
+        }
+    }
+}
+
+/// Represents a zome function declaration
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Hash)]
+pub struct FnDeclaration {
+    /// The name of this fn declaration.
+    #[serde(default)]
+    pub name: String,
+    pub inputs: Vec<FnParameter>,
+    pub outputs: Vec<FnParameter>,
+}
+
+impl Default for FnDeclaration {
+    /// Defaults for a "fn_declarations" object.
+    fn default() -> Self {
+        FnDeclaration {
+            name: String::new(),
+            inputs: Vec::new(),
+            outputs: Vec::new(),
+        }
+    }
+}
+
+impl FnDeclaration {
+    /// Allow sane defaults for `FnDecrlaration::new()`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+/// Represents an trait definition for bridging
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Hash)]
+pub struct Trait {
```

Comment:
So here's my understanding from the code: `Trait` is a lot like a `Capability`, but the only difference is that `Trait` contains the full function signature because it's referring to functions that might exist in other DNAs, whereas `Capability` is entirely inward-facing, and only holds the function names because it can just look up the function signatures from its own DNA. Is that right?

Reply:
Yep...  Trait is what's used at the composition level of the container for bridging.  Maybe we use a different name for it:  https://realtimeboard.com/app/board/o9J_kyiXmFs=/?moveToWidget=3074457346285220939